rust/src/sans_io/read.rs

```diff
@@ -0,0 +1,1014 @@
+                        std::mem::swap(&mut from, &mut self.from);
+                        let state = match &mut from {
+                            Chunk(state) => state,
+                            File => panic!(
+                                "invariant: padding should only be Some if reading from chunk"
+                            ),
+                        };
+                        if let Some(decompressor) = state.decompressor.take() {
+                            self.return_decompressor(decompressor)?
+                        }
+                        if let Some(hasher) = state.hasher.take() {
+                            let calculated = hasher.finalize();
+                            if state.crc != 0 && calculated != state.crc {
+                                self.failed = true;
+                                return Err(McapError::BadChunkCrc {
+                                    saved: state.crc,
+                                    calculated,
+                                });
+                            }
+                        }
+                        self.currently_reading = PaddingAfterChunk { len: padding }
+                    }
+                    return Ok(Some(ReadAction::GetRecord {
+                        data: &self.uncompressed.view(span)[9..],
+                        opcode,
+                    }));
+                }
+                CurrentlyReading::ChunkHeader { len } => {
+                    // Need to read _only_ the chunk header, which is of variable size.
+                    // We load the minimum chunk header size, which is always enough to check
+                    // the length of the compression string. With the compression string length,
+                    // the true length of the chunk header is known, so we read that.
+                    let min_chunk_header_len: usize = 8 + 8 + 8 + 4 + 4 + 8;
+                    if len < min_chunk_header_len as u64 {
+                        self.failed = true;
+                        return Err(McapError::RecordTooShort {
+                            opcode: op::CHUNK,
+                            len,
+                            expected: min_chunk_header_len as u64,
+                        });
+                    }
+                    let span = match self.load(min_chunk_header_len)? {
+                        Span(span) => span,
+                        Remainder(remainder) => return self.request(remainder),
+                    };
+                    let input = self.uncompressed.view(span);
+                    let compression_string_length =
+                        u32::from_le_bytes(input[28..32].try_into().unwrap());
+                    let true_chunk_header_len =
+                        min_chunk_header_len + compression_string_length as usize;
+                    if len < true_chunk_header_len as u64 {
+                        self.failed = true;
+                        return Err(McapError::RecordTooShort {
```

Comment:
not quite following the flow here - why doesn't this `len < true_chunk_header_len` turn into a `request()`? (if the bytes read so far were enough to determine the `compression_string_length` but not enough to actually read the full chunk header?)

Reply:
`len` is the record length that we got from the file right after the chunk opcode. If that length is shorter than`true_chunk_header_len`, that means the declared record length is too short to contain all the fields for a valid chunk record. If we request more bytes at this point, we'll start reading into the next record.